### ISSUE TITLE
Return to verifier service after proof request is complete

### ISSUE THREAD

**Issue by knguyenBC** [bcgov/bc-wallet-mobile#2181]
### Description of feature / user story
As a holder, I want to be automatically redirected to the service page after completing a proof request on the same device, so that I'm not lost or confused as to where to go next.

With the onset of online DAV, we need to hone in on the single device UX. 

### Acceptance Criteria
- On a single device experience, return the user to the service page that that sent the proof request

Given the user has received a proof request
And there's a goal code to inform user that the interaction is over
Then display the ["back-to-service" screens ](https://www.figma.com/design/mfUqAvZIeOsgxPqWTTp0SU/BC-Wallet-app?node-id=4101-2286&t=3vS9KnAnPrYv3Nxh-1)(iOS or Android specific)

### Wireframes or relevant image assets / links
User flow update with goal code : https://miro.com/app/board/uXjVNoMp-VI=/?moveToWidget=3458764618001255799&cot=14

Wireframes for Android and iOS specific back-to-service screens: https://www.figma.com/design/mfUqAvZIeOsgxPqWTTp0SU/BC-Wallet-app?node-id=4101-2286&t=3vS9KnAnPrYv3Nxh-1

**Comment by knguyenBC** [comment]
Talked with Emiliano and @cvarjao . @esune  has added a goal code to vc-authn proof requests so that BC Wallet could know to close the connection. Emiliano to research on how he can add a hint on which browser the user is using to add to payload and wallet to use to send the user to the right browser/app.

Spoke with @swcurran  and the current conclusion for multi-interaction
- We COULD use goal codes to signify a sequence of interactions (connection, verify, issue, message) etc.
- It would be complicated in that there would need some format and governance on the creation of these goal codes and how each wallet would read them. ex. aries.connection.5verifies vs. aries.connection.verifyx5
- it would be complicated to deal with errors in between complex flows. Ex. and error occurs after the first verification in a sequence.
- Action menus could also be used however the UX flow would not be seamless. It would work similar to being triaged in a automated call-center.

**Comment by swcurran** [comment]
Sorry I didn’t have a lot to contribute last week — I was on vacation.  

The trade off that Kim and I talked about was in the creation of use case specific goal codes for complicated workflows (e.g. a goal code that covers exactly the entire IAS connect/issue process for the person credential) vs. generic goal codes that are re-usable across a set of workflows.  The latter are preferred as they can be used in multiple use cases, but how do we balance goal codes semantics to cover the specific use cases we have.

I’m thinking the best way to do this might be to create a set of goal codes that cover only the "next step" after the current request — perhaps with both a “pass” and “fail” designation.  Goal Codes can be applied to an Invitation (Connect request), a credential offer and issue, and a presentation request.  So we could have goal codes like:

- On an invitation:
   - “Presentation” to indicate that after the connection is established a presentation request will follow.
   - “Issue” to indicate that after the connection is established an issue credential flow will follow.
   - “Complete” to indicate that nothing will happen after the connection is established.
- On an presentation:
   - “Presentation” to indicate that after the presentation completes, another presentation request will follow.
   - “Issue” to indicate that after after the presentation completes, an issue credential flow will follow.
   - “Complete” to indicate that nothing will happen after after the presentation completes.
- On an issue:
   - “Presentation” to indicate that after the issuance completes, a presentation request will follow.
   - “Issue” to indicate that after after the issuance completes, another issue credential flow will follow.
   - “Complete” to indicate that nothing will happen after after the issuance completes.

Perhaps the goal code could also have a “<action>-fail-<action>” to provide both a follow up to a “success” and “failure”.  And of course, there can be other options that could happen after an operation.

One question I wasn’t sure of. Kim mentioned that for the single-screen / all-mobile flow, there is the need to return to the invoking application.  I’m a correct in assuming that the “invoking application” is known to the Wallet, and the connected agent does NOT have to tell the Wallet what app to return to?

**Comment by esune** [comment]
For more complex interactions, Clecio suggested maybe it would make sense to use DRPC since the integration between the BC Wallet and the agent would be pretty tight anyway (i.e.: IAS) vs. coming up with complex goal codes - thoughts on this? It could be as simple as having an initial goal code telling BC Wallet `wait-for-end` and sending `end` once done.

Regarding the invoking application: it sounds like the wallet does not know which browser app on the device may have triggered the deep-link, so we will have to do some detection and add ad-hoc parameters for that return action to be feasible.

**Comment by swcurran** [comment]
DRPC could be used — good idea.  That would be very much along the “custom solution” side as the “frameworks” would just be passing payloads, with the controllers doing the work.

If the wallet doesn’t know who called it, wouldn’t users just use the link in the top right (e.g., “<— Safari”) to go back to the calling app?  That’s what the OS has been designed to encourage.

**Comment by esune** [comment]
Yes, we either need very explicit UX/instructions to get the users to do that, or we need a "hint" the app can use to automate the process.

Side note on goal codes: as far as I can tell BC Wallet/Bifold only parse and handle [some goal codes](https://github.com/openwallet-foundation/bifold-wallet/blob/51c2a678274eb6279b36baac2060f44b60577e0c/packages/legacy/core/App/screens/Connection.tsx#L46-L50) when handling an oob connection request, so additional handling would be required on the wallet end to support goal codes on existing connections, as well as to process custom-defined goal codes.

ACA-Py also appears to not have an explicit way of setting goal codes for `present-proof` and `issue-credential` protocols (based on the API), so further investigation and possibly tweaks will be required.

**Comment by Rousseau-Sylvain** [comment]
For proof request flows that are fully carried out on a mobile device, we needed a mechanism to redirect users to the appropriate flow when accepting the information-sharing request.

As part of the "WALK" approach in our solution, we opted for a simple implementation. We simply pass the necessary redirection information in the parameters of the invitationURL. 

This way, we obtain a proof request button (since we are on mobile) in the following format:
"< Deeplink: didcomm >://< shortenURL >?redirect= < Redirection Deeplink >"

We will add a button in the wallet to return to the browser only in the case of a fully mobile flow.
Here is an attached video showing the flow from the browser to the wallet and back to the browser.

https://github.com/user-attachments/assets/63be70a0-1dc7-42c2-8db5-f143ac1cf9cd

**Comment by esune** [comment]
@Rousseau-Sylvain what format/information are you adding to the redirection deeplink - would you be able to provide an example? We ave been looking into adding a hint about which web browser was used to initiate the interaction ( see https://github.com/openwallet-foundation/acapy-vc-authn-oidc/issues/731), however the deeplink option seems like it might potentially provide extra benefits?

**Comment by Rousseau-Sylvain** [comment]
@esune The service that initiates the verification determines if the application (the browser) the user is using is known. If it is, it simply adds a query parameter to the invitation with the deeplink to return to the application (the browser).

Example:

- The user is using Google Chrome.
- The browser is known ✅.
- Adding the query parameter ?redirect=googlechrome://.
- Once redirected to the wallet, the user sees the proofRequest, and if the user agrees to share, since the redirect parameter is known, we display the button that navigates to googlechrome://.

**Comment by loneil** [comment]
> * The browser is known ✅.

@Rousseau-Sylvain curious what is being used for the app deep link detection that gets passed in the URL?

Is the calling verification page using the browser user agent and mapping that to known deep link prefixes? And is there a supported browser list then where if I use an unsupported browser the wallet does not know which prefix? I know on my Android for example if I use the default Google app (search for the web service on my phone home screen for example) the user agent will say I'm in Chrome so I wonder if then the deeplink would not know about the page actually being in the Google app to return to rather than the Chrome app.

**Comment by esune** [comment]
> > * The browser is known ✅.
> 
> [@Rousseau-Sylvain](https://github.com/Rousseau-Sylvain) curious what is being used for the app deep link detection that gets passed in the URL?

+1 from me. In particular on whether the deep link URI scheme is something that is handled just for this scenario by the app, or is being extracted/determined using some official codified values (possibly the namespace registered by the app).
